go/worker/beacon/worker.go

```diff
@@ -0,0 +1,653 @@
+		w.logger.Error("round state mismatch",
+			"state", w.state.State,
+		)
+		return false
+	}
+
+	return true
+}
+
+func (w *Worker) onStateCommit(ev *beacon.SCRAPEEvent, scrapeState *beacon.SCRAPEState) {
+	if w.state != nil {
+		if ev.Epoch == w.state.Epoch && ev.Round == w.state.Round {
+			// This shouldn't happen, log something about it.
+			w.logger.Error("commit: received redundant round event, ignoring")
+			return
+		}
+		w.persistState(nil)
+	}
+
+	// New round.
+
+	// Determine if we are present in the participant list, and ensure
+	// that the point for each participant are valid.
+	var (
+		isPresent bool
+		points    []scrape.Point
+	)
+	for _, id := range ev.Participants {
+		if id.Equal(w.identity.NodeSigner.Public()) {
+			isPresent = true
+		}
+		node, err := w.registry.GetNode(
+			w.ctx,
+			&registry.IDQuery{
+				Height: ev.Height,
+				ID:     id,
+			},
+		)
+		if err != nil {
+			w.logger.Error("commit: failed to query node descriptor",
+				"err", err,
+				"id", id,
+			)
+			return
+		}
+		if node.Beacon == nil {
+			w.logger.Error("commit: participant node missing point",
+				"id", id,
+			)
+			return
+		}
+		points = append(points, node.Beacon.Point)
+	}
+	h, cmpH := hash.NewFrom(points), hash.NewFrom(scrapeState.Instance.Participants)
+	if !h.Equal(&cmpH) {
+		w.logger.Error("commit: participant point list mismatch")
+		return
+	}
+	if !isPresent {
+		w.logger.Info("commit: node is not present in participant list")
+		return
+	}
+
+	// Initialize the local SCRAPE instance.
+	instance, err := scrape.New(&scrape.Config{
+		PrivateKey:    &w.identity.BeaconScalar,
+		Participants:  points,
+		Threshold:     scrapeState.Instance.Threshold,
+		PVSSThreshold: scrapeState.Instance.PVSSThreshold,
+	})
+	if err != nil {
+		w.logger.Error("commit: failed to initialize SCRAPE instance",
+			"err", err,
+		)
+		return
+	}
+
+	// Generate the commit.
+	commit, err := instance.Commit()
+	if err != nil {
+		w.logger.Error("commit: failed to generate commit",
+			"err", err,
+		)
+		return
+	}
+	commitPayload := beacon.SCRAPECommit{
+		Epoch:  ev.Epoch,
+		Round:  ev.Round,
+		Commit: commit,
+	}
+	tx := transaction.NewTransaction(0, nil, beacon.MethodSCRAPECommit, commitPayload)
+
+	// Persist so it is possible to recover.
+	w.persistState(&roundState{
+		Epoch:    ev.Epoch,
+		Round:    ev.Round,
+		State:    ev.State,
+		Instance: instance,
+		CommitTx: tx,
+	})
+
+	// Submit the commit tx.
+	w.retrySubmitTx(tx)
+}
+
+func (w *Worker) onStateReveal(ev *beacon.SCRAPEEvent, scrapeState *beacon.SCRAPEState) {
+	var stateOk bool
+	defer func() {
+		if !stateOk {
+			w.persistState(nil)
+		}
+	}()
+
+	if !w.eventOkForState(ev, beacon.StateCommit) {
+		return
+	}
+
+	if scrapeState.BadParticipants[w.identity.NodeSigner.Public()] {
+		w.logger.Error("reveal: node identity listed as bad participant")
+		return
+	}
+
+	// Handle all of the commits.
+	//
+	// Note: This could be better optimized by having the consensus layer
```

Comment:
Maybe (once ready for merge) file a follow-up issue for a possible optimization.